### PR TITLE
Polly V8: Public API Review (Finalization)

### PR DESCRIPTION
**Pull request by martintmk**
### Details on the issue fix or feature implementation

The follow-up of #1233 to finalize the API review.  Contributes to #1301.

I transferred remaining unresolved comments here.

### Confirm the following

- [ ]  I started this PR by branching from the head of the default branch
- [ ]  I have targeted the PR to merge into the default branch
- [ ]  I have included unit tests for the issue/feature
- [ ]  I have successfully run a local build

### CODE REVIEW

File: ApiReview/API.Polly.Core/NoDocs/Polly/Outcome.TResult.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+using System.Runtime.CompilerServices;
+using System.Runtime.ExceptionServices;
+
+namespace Polly;
+
+public readonly struct Outcome<TResult>
+{
+    public Exception? Exception { get; }
+    public TResult? Result { get; }
+    public void EnsureSuccess();
```
Comment (martintmk) [inline ApiReview/API.Polly.Core/NoDocs/Polly/Outcome.TResult.cs:13]: Andrey:
I would expect Result getter to throw an exception if there was an exception.
similar to how Task.Result throws a captured exception.
but is this method even needed?

Martin:
I think we should avoid throwing exceptions from Outcome as it is low-level API used in high-perf scenarios and in delegates.

Juraj:
This is similar pattern that's used for `HttpResponseMessage` to check for successful status code. Also Cosmos DB SDK has similar implementation - you either work with DTOs directly (higher level) and get the exceptions, or you work with `ResponseMessage` (lower level) but need to check for the result explicitly using `response.EnsureSuccessStatusCode()`.

File: ApiReview/API.Polly.Core/NoDocs/Polly.Timeout/TimeoutStrategyOptions.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+using System.ComponentModel.DataAnnotations;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+
+namespace Polly.Timeout;
+
+public class TimeoutStrategyOptions : ResilienceStrategyOptions
+{
+    [Range(typeof(TimeSpan), "00:00:01", "1.00:00:00")]
+    public TimeSpan Timeout { get; set; }
```
Comment (martintmk) [inline ApiReview/API.Polly.Core/NoDocs/Polly.Timeout/TimeoutStrategyOptions.cs:13]:
```suggestion
    public TimeSpan DefaultTimeout { get; set; }
```

File: ApiReview/API.Polly.RateLimiting/NoDocs/Polly.RateLimiting/RateLimiterStrategyOptions.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+using System.Threading.Tasks;
+
+namespace Polly.RateLimiting;
+
+public class RateLimiterStrategyOptions : ResilienceStrategyOptions
+{
+    [Required]
+    public ConcurrencyLimiterOptions DefaultRateLimiterOptions { get; set; }
+    public Func<OnRateLimiterRejectedArguments, ValueTask>? OnRejected { get; set; }
+    public ResilienceRateLimiter? RateLimiter { get; set; }
```
Comment (martintmk) [inline ApiReview/API.Polly.RateLimiting/NoDocs/Polly.RateLimiting/RateLimiterStrategyOptions.cs:16]:
```suggestion
   public Func<RateLimiterArguments, ValueTask<RateLimitLease>> RateLimiter { get; set; }
```

And get rid of `ResilienceRateLimiter`. We would have to solve the problem with resource management i.e. disposing unused rate limiters after pipeline is reloaded or disposed.

File: ApiReview/API.Polly.Core/NoDocs/Polly/OutcomeArguments.cs [hunk 4]
```diff
@@ -0,0 +1,16 @@
+// Assembly 'Polly.Core'
+
+using System;
+using System.Runtime.CompilerServices;
+
+namespace Polly;
+
+public readonly struct OutcomeArguments<TResult, TArgs>
```
Comment (martintmk) [inline ApiReview/API.Polly.Core/NoDocs/Polly/OutcomeArguments.cs:8]: Can we drop `OutcomeArguments` and move `Context` and `Outcome` to each respective arguments?

For example:

``` csharp
public readonly struct OnRetryArguments<TResult>
{
    // moved here
    public Outcome<TResult> Outcome { get; }

    // moved here
    public ResilienceContext Context { get; }

    public int AttemptNumber { get; }

    public TimeSpan RetryDelay { get; }

    public TimeSpan ExecutionTime { get; }
}
```

Now the delegate signature instead of:

``` csharp
    public Func<OutcomeArguments<TResult, OnRetryArguments>, ValueTask>? OnRetry { get; set; }
```

Becomes:

``` csharp
    public Func<OnRetryArguments<TResult>, ValueTask>? OnRetry { get; set; }
```

Upsides:

- Cleaner and easier to consume API
- Arguments are aligned with non-outcome based ones (`OnTimeoutArguments` for example)

Downsides:

- Duplication of properties across multiple argument classes.
- Arguments are now generic which makes them harder to consume.

File: ApiReview/API.Polly.Core/NoDocs/Polly/ResilienceProperties.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
+// Assembly 'Polly.Core'
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace Polly;
+
+public sealed class ResilienceProperties
```
Comment (martintmk) [inline ApiReview/API.Polly.Core/NoDocs/Polly/ResilienceProperties.cs:9]: Andrey:
can we mimic the behaviour of `IFeatureCollection` here by changing the API to use the `TValue` type as a key? this eliminates the need to generate custom names for properties and have the artificial `PropertyKey` type and fosters the everything is strong typed approach

```cs
public class ResilienceProperties
{
    void Set<TValue>(TValue? value);
    TValue? Get<TValue>();
}
```

Martin:
One common use-case of V7 is attaching primitive types (see #890 as an example). While in V8 they could use something like `properties.Set<int>(3)` or `properties.Set<MyIntWrapper>()` it could lead to ambiguity or more more complicated usage.

File: ApiReview/API.Polly.Testing/NoDocs/Polly.Testing/ResiliencePipelineDescriptor.cs [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+using System.Runtime.CompilerServices;
+
+namespace Polly.Testing;
+
+public sealed class ResiliencePipelineDescriptor
+{
+    public IReadOnlyList<ResilienceStrategyDescriptor> Strategies { get; }
+    public ResilienceStrategyDescriptor FirstStrategy { get; }
+    public bool IsReloadable { get; }
+    public ResiliencePipelineDescriptor(IReadOnlyList<ResilienceStrategyDescriptor> strategies, bool isReloadable);
```
Comment (martintmk) [inline ApiReview/API.Polly.Testing/NoDocs/Polly.Testing/ResiliencePipelineDescriptor.cs:13]: Hide this constructor.

File: ApiReview/API.Polly.Core/NoDocs/Polly.Retry/OnRetryArguments.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+using System;
+using System.Runtime.CompilerServices;
+
+namespace Polly.Retry;
+
+public readonly struct OnRetryArguments
+{
+    public int AttemptNumber { get; }
+    public TimeSpan RetryDelay { get; }
+    public TimeSpan ExecutionTime { get; }
```
Comment (martintmk) [inline ApiReview/API.Polly.Core/NoDocs/Polly.Retry/OnRetryArguments.cs:12]:
```suggestion
    public TimeSpan Duration { get; }
```

File: ApiReview/API.Polly.Core/NoDocs/Polly/Outcome.TResult.cs (hunk 1 above)
Comment (martincostello) [reply ApiReview/API.Polly.Core/NoDocs/Polly/Outcome.TResult.cs:13]: I prefer the HttpClient-style pattern too.

File: ApiReview/API.Polly.Core/NoDocs/Polly.Retry/OnRetryArguments.cs (hunk 7 above)
Comment (martintmk) [reply ApiReview/API.Polly.Core/NoDocs/Polly.Retry/OnRetryArguments.cs:12]: Addressed by #1508

File: ApiReview/API.Polly.Testing/NoDocs/Polly.Testing/ResiliencePipelineDescriptor.cs (hunk 6 above)
Comment (martintmk) [reply ApiReview/API.Polly.Testing/NoDocs/Polly.Testing/ResiliencePipelineDescriptor.cs:13]: Addressed by #1508

File: ApiReview/API.Polly.Core/NoDocs/Polly/ResilienceProperties.cs (hunk 5 above)
Comment (andrey-noskov) [reply ApiReview/API.Polly.Core/NoDocs/Polly/ResilienceProperties.cs:9]: can you elaborate on this a bit more
- the current approach implies that a caller must create an instance of PropertyKey to hold the key
- the suggested approach allows a caller to use custom types as well - making the values strong typed - but also primitives as you pointed out above

File: ApiReview/API.Polly.RateLimiting/NoDocs/Polly.RateLimiting/RateLimiterStrategyOptions.cs (hunk 3 above)
Comment (martintmk) [reply ApiReview/API.Polly.RateLimiting/NoDocs/Polly.RateLimiting/RateLimiterStrategyOptions.cs:16]: Addressed by #1509

File: ApiReview/API.Polly.Core/NoDocs/Polly/Outcome.TResult.cs (hunk 1 above)
Comment (martintmk) [reply ApiReview/API.Polly.Core/NoDocs/Polly/Outcome.TResult.cs:13]: Little bit more context:

- This type is semantically/behaviorally almost identical to [`DelegateResult`](https://github.com/App-vNext/Polly/blob/282297e95b6b6ac140f1c6240fd0b21924cbb6ee/src/Polly/DelegateResult.cs#L6) or [`PolicyResult`](https://github.com/App-vNext/Polly/blob/282297e95b6b6ac140f1c6240fd0b21924cbb6ee/src/Polly/PolicyResult.cs#L6) from V7. 
- The `EnsureSuccess` was added as a feature request by user (#1238). We didn't want to add any new API to V7 so we included this method in V8.

File: ApiReview/API.Polly.Core/NoDocs/Polly/ResilienceProperties.cs (hunk 5 above)
Comment (martintmk) [reply ApiReview/API.Polly.Core/NoDocs/Polly/ResilienceProperties.cs:9]: > the current approach implies that a caller must create an instance of PropertyKey to hold the key

Correct, we want to avoid magic strings or value casting. Key encapsulates both type and key to the value. 

> the suggested approach allows a caller to use custom types as well - making the values strong typed - but also primitives as you pointed out above

Yes, there might be some performance penalty if you want to store `string` for example as extra wrapper will be required to store the value. 

Another downside is unfamiliarity for Polly V7 users. Migrating from `Context` to `ResilienceProperties` is relatively straightforward. You don't even have to hold the key, you can just create it inline.

This API is also familiar for users that work with HttpClient:
- https://source.dot.net/#System.Net.Http/System/Net/Http/HttpRequestOptionsKey.cs,12

Or with RateLimiting:
- https://source.dot.net/#System.Threading.RateLimiting/System/Threading/RateLimiting/MetadataName.T.cs,12

Comment (andrey-noskov) [reply ApiReview/API.Polly.Core/NoDocs/Polly/ResilienceProperties.cs:9]: on the contrary, the typed approach must be familiar to all ASP.NET users

File: ApiReview/API.Polly.Core/NoDocs/Polly.Registry/ConfigureBuilderContext.cs [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace Polly.Registry;
+
+public class ConfigureBuilderContext<TKey> where TKey : notnull
+{
+    public TKey PipelineKey { get; }
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public void EnableReloads(Func<Func<CancellationToken>> tokenProducerFactory);
```
Comment (martintmk) [inline ApiReview/API.Polly.Core/NoDocs/Polly.Registry/ConfigureBuilderContext.cs:14]:
```suggestion
    public void AddReloadToken(CancellationToken cancellationToken);
```
Would simplify the API and allows to add multiple reload sources to the context.

File: ApiReview/API.Polly.Core/NoDocs/Polly/OutcomeArguments.cs (hunk 4 above)
Comment (martincostello) [reply ApiReview/API.Polly.Core/NoDocs/Polly/OutcomeArguments.cs:8]: Do you have an example of code consuming this for before and after this change? Would make it easier to make a judgement on the downsides for the consumer.

Comment (martintmk) [reply ApiReview/API.Polly.Core/NoDocs/Polly/OutcomeArguments.cs:8]: Sure, for delegates it would look like:

``` csharp

// before
new RetryStrategyOptions
{
    OnRetry = args =>
    {
        Console.WriteLine("Attempt: {0}", args.Arguments.AttemptNumber);
        return default;
    }
}

// after
new RetryStrategyOptions
{
    OnRetry = args =>
    {
        Console.WriteLine("Attempt: {0}", args.AttemptNumber);
        return default;
    }
}
```

For telemetry:

``` csharp
//before
private class CustomTelemetryListener : TelemetryListener
{
    public override void Write<TResult, TArgs>(in TelemetryEventArguments<TResult, TArgs> args)
    {
        if (args.Arguments is OnRetryArguments retryArgs)
        {
            Console.WriteLine("Attempt: {0}", retryArgs.AttemptNumber);
        }
    }
}


// after
private class Listener : TelemetryListener
{
    public override void Write<TResult, TArgs>(in TelemetryEventArguments<TResult, TArgs> args)
    {
        if (args.Arguments is OnRetryArguments<TResult> retryArgs)
        {
            Console.WriteLine("Attempt: {0}", retryArgs.AttemptNumber);
        }
    }
}
```

Major use-case are the delegates where it looks simpler. 
For listeners it would be a problem before where the `Write` method wasn't generic. Now it's doable.

Comment (martincostello) [reply ApiReview/API.Polly.Core/NoDocs/Polly/OutcomeArguments.cs:8]: I think the proposed change looks good then.

Comment (vany0114) [reply ApiReview/API.Polly.Core/NoDocs/Polly/OutcomeArguments.cs:8]: @martintmk I guess not all the arguments are gonna be generic, only the ones for reactive strategies right? 

```c#
public readonly struct OnRetryArguments<TResult>
```

And yeah from the simplicity standpoint, I lean towards removing the `OutcomeArguments` envelope, actually, that's how I initially used it in Simmy for reactive chaos strategies.

Comment (martintmk) [reply ApiReview/API.Polly.Core/NoDocs/Polly/OutcomeArguments.cs:8]: >  I guess not all the arguments are gonna be generic, only the ones for reactive strategies right?

Yes, reactive strategies only.

Thanks folks, I'll prepare the PR with changes.

Comment (martintmk) [reply ApiReview/API.Polly.Core/NoDocs/Polly/OutcomeArguments.cs]: Addressed by #1513

File: ApiReview/API.Polly.Core/NoDocs/Polly.Hedging/OnHedgingArguments.cs [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+using Polly.Utils;
+
+namespace Polly.Hedging;
+
+public readonly struct OnHedgingArguments<TResult>
+{
+    public Outcome<TResult> Outcome { get; }
+    public ResilienceContext Context { get; }
+    public int AttemptNumber { get; }
+    public bool HasOutcome { get; }
```
Comment (martintmk) [inline ApiReview/API.Polly.Core/NoDocs/Polly.Hedging/OnHedgingArguments.cs:14]: This tells the consumer that no outcome was produced in previous attempt due to operation taking too long. 

Maybe find a better name or just make `Outcome` property nullable (since now we have it).

File: ApiReview/API.Polly.Core/NoDocs/Polly.Fallback/FallbackStrategyOptions.cs [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+using System.Threading.Tasks;
+
+namespace Polly.Fallback;
+
+public class FallbackStrategyOptions<TResult> : ResilienceStrategyOptions
+{
+    [Required]
+    public Func<FallbackPredicateArguments<TResult>, ValueTask<bool>> ShouldHandle { get; set; }
+    [Required]
+    public Func<FallbackPredicateArguments<TResult>, ValueTask<Outcome<TResult>>>? FallbackAction { get; set; }
```
Comment (martintmk) [inline ApiReview/API.Polly.Core/NoDocs/Polly.Fallback/FallbackStrategyOptions.cs:15]:
```suggestion
    public Func<FallbackActionArguments<TResult>, ValueTask<Outcome<TResult>>>? FallbackAction { get; set; }
```

File: ApiReview/API.Polly.Core/NoDocs/Polly.Timeout/TimeoutStrategyOptions.cs (hunk 2 above)
Comment (martintmk) [reply ApiReview/API.Polly.Core/NoDocs/Polly.Timeout/TimeoutStrategyOptions.cs:13]: In hedging we have:

- `HedgingDelay`
- `HedgingDelayGenerator`

It's very similar i timeout too:

- `Timeout`
- `TimeoutGenerator`

I think it's ok to keep the simpler `Timeout` name, otherwise for consistency we should also use `DefaultHedgingDelay`, which imho is too wordy.

File: ApiReview/API.Polly.Core/NoDocs/Polly.Retry/RetryDelayGeneratorArguments.cs [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+using System.Runtime.CompilerServices;
+
+namespace Polly.Retry;
+
+public readonly struct RetryDelayGeneratorArguments<TResult>
+{
+    public Outcome<TResult> Outcome { get; }
+    public ResilienceContext Context { get; }
+    public int AttemptNumber { get; }
+    public TimeSpan DelayHint { get; }
```
Comment (andrey-noskov) [inline ApiReview/API.Polly.Core/NoDocs/Polly.Retry/RetryDelayGeneratorArguments.cs:13]: what is the purpose of this hint?

File: ApiReview/API.Polly.Core/NoDocs/Polly.Hedging/HedgingStrategyOptions.cs [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+using System;
+using System.ComponentModel.DataAnnotations;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+
+namespace Polly.Hedging;
+
+public class HedgingStrategyOptions<TResult> : ResilienceStrategyOptions
+{
+    public TimeSpan HedgingDelay { get; set; }
```
Comment (juraj-blazek) [inline ApiReview/API.Polly.Core/NoDocs/Polly.Hedging/HedgingStrategyOptions.cs:12]: Use `Default` prefix for the default delay/timeout properties.

File: ApiReview/API.Polly.Core/NoDocs/Polly.Retry/RetryStrategyOptions.TResult.cs [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+namespace Polly.Retry;
+
+public class RetryStrategyOptions<TResult> : ResilienceStrategyOptions
+{
+    [Range(1, int.MaxValue)]
+    public int RetryCount { get; set; }
+    public RetryBackoffType BackoffType { get; set; }
+    public bool UseJitter { get; set; }
+    [Range(typeof(TimeSpan), "00:00:00", "1.00:00:00")]
+    public TimeSpan BaseDelay { get; set; }
```
Comment (juraj-blazek) [inline ApiReview/API.Polly.Core/NoDocs/Polly.Retry/RetryStrategyOptions.TResult.cs:18]: `BaseDelay` -> `Delay`

File: ApiReview/API.Polly.Core/NoDocs/Polly/Outcome.TResult.cs (hunk 1 above)
Comment (martintmk) [reply ApiReview/API.Polly.Core/NoDocs/Polly/Outcome.TResult.cs:13]: How do you think about renaming `EnsureSuccess` to `ThrowIfException`?

File: ApiReview/API.Polly.Core/NoDocs/Polly.Retry/RetryDelayGeneratorArguments.cs (hunk 11 above)
Comment (martintmk) [reply ApiReview/API.Polly.Core/NoDocs/Polly.Retry/RetryDelayGeneratorArguments.cs:13]: This value is generated by the retry strategy and respects the the retry options and current attempt number.
If generator is not able to extract the delay from the  outcome it can return this value as a next delay attempt.

File: ApiReview/API.Polly.Core/NoDocs/Polly/ResilienceProperties.cs (hunk 5 above)
Comment (martincostello) [reply ApiReview/API.Polly.Core/NoDocs/Polly/ResilienceProperties.cs:9]: I don't follow the suggestion here. Can you give some examples in code of usage with the current code and with the proposed change?

IFeatureCollection is typically 1:1 is it not, with there only ever being a single instance of a given type representing a concept? It seems to me that this suggestion would make it harder to attach multiple instances of a specific type to the context, such as strings where you might want multiple types of IDs associated with it.

Comment (andrey-noskov) [reply ApiReview/API.Polly.Core/NoDocs/Polly/ResilienceProperties.cs:9]: ASP.NET Core's typed approach requires users to recognize that every entity passed through a pipeline via the context object represents a contract between two or more components, and as such, it'd better be well-defined, provide encapsulation, be future-proof, and so on. 

imo, the current approach, on the other hand, encourages the usage of the primitive obsession pattern by focusing solely on the keys rather than te entities.

> you might want multiple types of IDs

every ID in this example has some role, some rules and agreements around it, may be every ID must be encoded according to some format, there is a set of operations that can be performed or even things like what to do when it isn't presented. all these aspects can't be transferred over with a mere string entity. instead, the approach chosen by the ASP.NET Core team forces one to use the OOP principles to encode the contract by creating a strong type for the entities and implementing the contract within the new type. For example

```cs
public class MyIDs
{
     string MyFirstId { get; init; }
     string? MySeconfId { get; init; }
}

// publish
var ids = new MyIDs(..);
ResilienceProperties.Set(ids);

// retrieve
var myIds= ResilienceProperties.Get<MyIDs>();
```

if the ids need to be modifiable, one can reflect that via the API of the type. if some of the ids are optiona, make the properties nullable. if the ids aren't simple arbitrary strings, add the validation to the ctor.

File: ApiReview/API.Polly.Core/NoDocs/Polly/Outcome.cs [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+
+using System;
+using System.Threading.Tasks;
+
+namespace Polly;
+
+public static class Outcome
+{
+    public static Outcome<TResult> FromResult<TResult>(TResult? value);
+    public static ValueTask<Outcome<TResult>> FromResultAsTask<TResult>(TResult value);
```
Comment (andrey-noskov) [inline ApiReview/API.Polly.Core/NoDocs/Polly/Outcome.cs:11]: while I find this helper useful, I believe if replace it with an extension for Outcome, it will enable more use-cases

`var task = Outcome.FromResult("dummy").AsTask()`

File: ApiReview/API.Polly.Core/NoDocs/Polly/Outcome.TResult.cs (hunk 1 above)
Comment (andrey-noskov) [reply ApiReview/API.Polly.Core/NoDocs/Polly/Outcome.TResult.cs:13]: (as I suggested this change) I believe this change will make the contract cleaner: (1) the Throw part  informs that the method can throw an exception and (2) the IfException explain the condition under which it can be thrown.

File: ApiReview/API.Polly.Core/NoDocs/Polly/PredicateResult.cs [hunk 15]
```diff
@@ -0,0 +1,11 @@
+// Assembly 'Polly.Core'
+
+using System.Threading.Tasks;
+
+namespace Polly;
+
+public static class PredicateResult
```
Comment (andrey-noskov) [inline ApiReview/API.Polly.Core/NoDocs/Polly/PredicateResult.cs:7]: from what I understand this is a factory like type which returns new instances on every call - because they are ValueTasks

Returning new instances from getter is considered bad practice.
I quickly googled to prove my point and found this gorgeous explanation https://stackoverflow.com/questions/2101646/is-object-creation-in-getters-bad-practice. see the accepted answer explaining the best practices compiled from a couple of well known champions

my proposal is to change the properties to methods, ie. `PredicateResult.True()`

File: ApiReview/API.Polly.Core/NoDocs/Polly.CircuitBreaker/BrokenCircuitException.TResult.cs [hunk 16]
```diff
@@ -0,0 +1,14 @@
+// Assembly 'Polly.Core'
+
+using System;
+using System.Runtime.CompilerServices;
+
+namespace Polly.CircuitBreaker;
+
+public class BrokenCircuitException<TResult> : BrokenCircuitException
+{
+    public TResult Result { get; }
```
Comment (andrey-noskov) [inline ApiReview/API.Polly.Core/NoDocs/Polly.CircuitBreaker/BrokenCircuitException.TResult.cs:10]: what is the role of this result? if a CB is open, there is no result, right?

File: ApiReview/API.Polly.Core/NoDocs/Polly.CircuitBreaker/OnCircuitOpenedArguments.cs [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+using Polly.Utils;
+
+namespace Polly.CircuitBreaker;
+
+public readonly struct OnCircuitOpenedArguments<TResult>
+{
+    public Outcome<TResult> Outcome { get; }
+    public ResilienceContext Context { get; }
+    public TimeSpan BreakDuration { get; }
+    public bool IsManual { get; }
```
Comment (andrey-noskov) [inline ApiReview/API.Polly.Core/NoDocs/Polly.CircuitBreaker/OnCircuitOpenedArguments.cs:14]: ``suggestion
    public bool IsIsolated { get; }
```

File: ApiReview/API.Polly.Core/NoDocs/Polly.CircuitBreaker/BrokenCircuitException.TResult.cs (hunk 16 above)
Comment (martintmk) [reply ApiReview/API.Polly.Core/NoDocs/Polly.CircuitBreaker/BrokenCircuitException.TResult.cs:10]: It returns result that opened the circuit. It's V7 API and we need to keep it.

File: ApiReview/API.Polly.Core/NoDocs/Polly/Outcome.cs (hunk 14 above)
Comment (martintmk) [reply ApiReview/API.Polly.Core/NoDocs/Polly/Outcome.cs:11]: See #1328 where we discuss this API (and `AsValueTask` as well). 

The following comment exaplains why we stuck with `FromResultAsTask`:

https://github.com/App-vNext/Polly/issues/1328#issuecomment-1598257460

File: ApiReview/API.Polly.Core/NoDocs/Polly/PredicateResult.cs (hunk 15 above)
Comment (martincostello) [reply ApiReview/API.Polly.Core/NoDocs/Polly/PredicateResult.cs:7]: I don't use ValueTask often so maybe there's something I'm missing that means this won't work, but why don't we just return one we've already created?

```csharp
public static class PredicateResult
{
    public static ValueTask<bool> True { get; } = new(true);
    public static ValueTask<bool> False { get; } = new(false);
}
```

Comment (martincostello) [reply ApiReview/API.Polly.Core/NoDocs/Polly/PredicateResult.cs:7]: OK, then yes in that case let's change it to a method.

File: ApiReview/API.Polly.Core/NoDocs/Polly/ResilienceProperties.cs (hunk 5 above)
Comment (juraj-blazek) [reply ApiReview/API.Polly.Core/NoDocs/Polly/ResilienceProperties.cs:9]: Keeping as is.

Comment (andrey-noskov) [reply ApiReview/API.Polly.Core/NoDocs/Polly/ResilienceProperties.cs:9]: fair enough

File: ApiReview/API.Polly.Core/NoDocs/Polly/Outcome.cs (hunk 14 above)
Comment (juraj-blazek) [reply ApiReview/API.Polly.Core/NoDocs/Polly/Outcome.cs:11]: Consider renaming to `FromResultAsValueTask` to make clear what it returns

File: ApiReview/API.Polly.Core/NoDocs/Polly.CircuitBreaker/BrokenCircuitException.TResult.cs (hunk 16 above)
Comment (juraj-blazek) [reply ApiReview/API.Polly.Core/NoDocs/Polly.CircuitBreaker/BrokenCircuitException.TResult.cs:10]: Let's move this back to Polly and v8 will only throw `BrokenCircuitException`
